tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -12626,26 +12627,49 @@ HLSLWaveSizeAttr *ValidateWaveSizeAttributes(Sema &S, Decl *D,
       A.getAttributeSpellingListIndex());
 
   pAttr->setSpelledArgsCount(A.getNumArgs());
-  int minWave = pAttr->getMin();
-  int maxWave = pAttr->getMax();
-  int prefWave = pAttr->getPreferred();
 
-  // validate attribute arguments.
-  if (!DXIL::IsValidWaveSizeValue(minWave, maxWave, prefWave)) {
-    S.Diag(A.getLoc(), diag::err_hlsl_wavesize_size)
-        << DXIL::kMinWaveSize << DXIL::kMaxWaveSize;
-  }
```

Comment:
I have a strong preference for the way this code was before. The way it is written now I find much harder to read. I understand that this reduces duplicated logic, but it is at the expense of readability.

Feel free to ignore this feedback because I don't feel strongly about it.

Reply:
This catches and diagnoses things in a less confusing way than before.  For instance, the previous code would not catch the max==0 case.  And in that case, if preferred was non-zero, the error it would give is that it isn't in range [min..0].  Reviewing the previous code to ensure it catches all the cases correctly was actually harder, because some cases would be blocked almost as a side effect of other things, such as the max==0 cases.

The new approach could also improve the diagnostics a bit without changing the validation logic.  For instance by explicitly identifying degenerate cases that should have been written in different syntax in HLSL, which isn't even hinted at by the diagnostics we have now.  The previous logic collapsed cases like this and gives the confusing diagnostic that min must be less than max, when min==max, instead of saying you must omit the max parameter to indicate this case.  Similar issue with preferred==0.

I didn't change these diagnostics, but they could be improved in the future based on the more specific validation failures available now.  Plus the explicit NumArgs checks could be changed to emit more useful diagnostics, in which case, centralizing these checks in DXILWaveSize::Validate would make even less sense.
 
Here are some cases I notice that the prior code didn't handle well (might not be complete).  These weren't caught in review because it wasn't as straightforward as it looked.
- Sema: wouldn't catch when max is set to 0 explicitly
- DxilValidation: wouldn't catch invalid preferred when max is 0
- DxilValidation: incorrect validation error when max is invalid (would complain about Min)
- DxilValidation: incorrect validation error when preferred is invalid but in range (would complain about Min)

Reply:
One more I just thought of:
- Sema: wouldn't catch when preferred is set to 0 explicitly.